CHANGELOG.md

```diff
@@ -0,0 +1,137 @@
+- New `type_alias::Map` defined as `rustc_hash::FxHashMap`.
+- New `type_alias::SelectedDependencies<P, V>` defined as `Map<P, V>`.
+- The types `Dependencies` and `DependencyConstraints` were introduced to clarify intent.
+- New function `choose_package_with_fewest_versions` to help implement
+  the `choose_package_version` method of a `DependencyProvider`.
+- Implement `FromStr` for `SemanticVersion`.
+- Add the `VersionParseError` type for parsing of semantic versions.
+
+#### Changed
+
+- Using SPDX license identifiers instead of MPL-2.0 classic file headers.
+- `./github/workflows/` CI workflow was improved, including check for conventional commits.
+- The `Solver` trait was replaced by a `DependencyProvider` trait
+  which now must implement a `choose_package_version` method
+  instead of `list_available_versions`.
+  So it now has the ability to choose a package in addition to a version.
+  The `DependencyProvider` also has a new optional method `should_cancel`
+  that may be used to stop the solver if needed.
+- The `choose_package_version` and `get_dependencies` methods of the
+  `DependencyProvider` trait now take an immutable reference to `self`.
+  Interior mutability can be used by implementor if mutability is needed.
+- The `Solver.run` method was thus replaced by a free function `solver::resolve`
+  taking a dependency provider as first argument.
+- The `OfflineSolver` is thus replaced by an `OfflineDependencyProvider`.
+- `SemanticVersion` now takes `u32` instead of `usize` for its 3 parts.
+- `NumberVersion` now uses `u32` instead of `usize`.
+
+#### Removed
+
+- `ErrorRetrievingVersions` variant of `error::PubGrubError`.
+
+### Changes in the internal parts of the API
+
+#### Added
+
+- `benches/large_case.rs` enables benchmarking of serialized registries of packages.
+- `examples/caching_dependency_provider.rs` an example dependency provider caching dependencies.
+- `PackageTerm<P, V> = (P, Term<V>)` new type alias for readability.
+- `Memory.term_intersection_for_package(&mut self, package: &P) -> Option<&Term<V>>`
+- New types were introduces for conflict resolution in `internal::partial_solution`
+  to clarify the intent and return values of some functions.
+  Those types are `DatedAssignment` and `SatisfierAndPreviousHistory`.
+- `PartialSolution.term_intersection_for_package` calling the same function
+  from its `memory`.
+- New property tests for ranges: `negate_contains_opposite`, `intesection_contains_both`
+  and `union_contains_either`.
+- A large synthetic test case was added in `test-examples/`.
+- A new test example `double_choices` was added
+  for the detection of a bug (fixed) in the implementation.
+- Property testing of big synthetic datasets was added in `tests/proptest.rs`.
+- Comparison of PubGrub solver and a SAT solver
+  was added with `tests/sat_dependency_provider.rs`.
+- Other regression and unit tests were added to `tests/tests.rs`.
+
+#### Changed
+
```

Comment:
```suggestion

- CI workflow was improved (`./github/workflows/`), including a check for [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and [Clippy ](https://github.com/rust-lang/rust-clippy) for source code linting.
- Using SPDX license identifiers instead of MPL-2.0 classic file headers.
```